api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Proxy-WASM ABI.
+ */
+// NOLINT(namespace-envoy)
+
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+//
+//  ABI calls from the host into the VM.
+//
+
+// Non-stream calls.
+
+/**
+ * Called when the VM starts for each plugin.
+ * @param root_context_id is an identifier with the lifetime of the VM which will be used for
+ * one or more related plugins and their corresponding proxy_on_configure(), proxy_on_done() and
+ * proxy_on_delete() calls. It is also be provided during creation of stream context for those
+ * plugins.
+ * @param configuration_size is the size of any configuration available via proxy_get_configuration.
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+extern "C" uint32_t proxy_on_start(uint32_t root_context_id, uint32_t configuration_size);
+/**
+ * Can be called to validate a configuration (e.g. by an xDS provider) both before proxy_on_start()
+ * to verify the VM configuration or after proxy_on_start() to verify a plugin configuation.
+ * @param root_context_id is a unique identifier for the configuration verification context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (i.e. bad configuration).
+ */
+extern "C" uint32_t proxy_validate_configuration(uint32_t root_context_id,
+                                                 uint32_t configuration_size);
+/**
+ * Called when a plugin loads or when plugin configuration changes dynamically.
```

Comment:
How does config change dynamically for a plugin? We reinstantiate the entire filter chain on each request, and config is immutable for the lifetime of a request.

Reply:
The VM has a lifetime which is longer than a request.  The VM is created when the bootstrap or xDS configuration is loaded, and is used for all requests over its lifetime (e.g. the lifetime of the Listener).  New configuration can be loaded to an existing VM as well. When a request starts at the discretion of the plugin author it can snap the current configuration inside the VM and when a new configuration arrives, it can continue to use the snapped configuration.  The ABI provides the facility, but how it is used is up to the plugin author.

Reply:
I think it would be clearer if we use identifiers such as `vm_config` and `plugin_config` consistently (modulo the bike shedding in the other PR over whether to call this plugin at all).

Reply:
Added the vm_ and plugin_ prefixes as required.   proxy_get_configuration and proxy_validate_configuration both work for either type of configuration so they are unqualified.